spec/cloud.go

```diff
@@ -148,12 +147,80 @@ func httpCli() *http.Client {
 // EC2Generator meta generator for EC2
 type EC2Generator struct {
 	baseURL *url.URL
+
+	mu       sync.Mutex
+	token    string
+	deadline time.Time
 }
 
 // IsEC2 checks current environment is EC2 or not
 func (g *EC2Generator) IsEC2(ctx context.Context) bool {
 	// implementation varies between OSs. see isec2_XXX.go
-	return isEC2(ctx)
+	return g.isEC2(ctx)
+}
+
+// check whether IMDS(Instance Metadata Service) is available.
+// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
+func (g *EC2Generator) hasMetadataService(ctx context.Context) (bool, error) {
+	cl := httpCli()
+
+	// try to refresh api token for IMDSv2
+	if token := g.refreshToken(ctx); token != "" {
+		return true, nil
```

Comment:
(And even in such cases, user can specify `cloud_platform="none"` to avoid the problem. ref: https://github.com/mackerelio/mackerel-agent/pull/447 )

Reply:
Should we check instance metadata more carefully?

Some cloud services provide Instance Metadata on 169.254.169.254 
https://gist.github.com/mrtc0/60ca6ba0fdfb4be0ba499c65932ab42e
And actually, there is wrong EC2 instance-id detection issue on another product e.g. https://github.com/fujiwara/AWS-XRay/issues/12

I think there are some other options.

- check http://169.254.169.254/latest/meta-data/services/partition 
  - ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
  - we know that it will return `aws` or `aws-partitionname`.
- check http://169.254.169.254/latest/dynamic/instance-identity/document
  - ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
  - The instance identity document is accompanied by a signature and a PKCS7 signature, so we can verify that it is signed by AWS.

Reply:
Thank you for suggestions.

First, (as [I've already commented above](https://github.com/mackerelio/mackerel-agent/pull/627#discussion_r370494322),) this pull request itself is okay, since the situation will not get worse.  And also, we already perform UUID check on [Linux](https://github.com/mackerelio/mackerel-agent/blob/master/spec/isec2_linux.go#L17) and [Windows](https://github.com/mackerelio/mackerel-agent/blob/master/spec/isec2_windows.go#L30), I believe current mackerel-agent has reasonable implementation.

But in future, if we need more precise detection, your first suggestion, verifying `/latest/meta-data/services/partition` can be a good solution.